backend/crossref.py

```diff
@@ -535,25 +537,49 @@ def fetch_all_records(self, filters=None,cursor="*"):
                 raise MetadataSourceException('Error while fetching CrossRef results:\nError was: '+str(e))
 
 
-    def fetch_and_save_new_records(self, starting_cursor='*'):
+    def fetch_and_save_new_records(self, starting_cursor='*', batch_time=datetime.timedelta(days=1)):
         """
         Fetches and stores all new Crossref records updated since the
         last update time of the associated OaiSource.
         """
         source = OaiSource.objects.get(identifier='crossref')
-        last_updated = source.last_update
 
-        to_update = self.fetch_all_records(filters={'from-update-date':last_updated.date().isoformat()},
-                cursor=starting_cursor)
-        for record in to_update:
-            try:
-                bare_paper = self.save_doi_metadata(record)
-                p = Paper.from_bare(bare_paper)
-                p.update_index()
-            except ValueError as e:
-                print(e)
+        # We substract one day to 'until-update-date' parameter as it is inclusive
+        one_day = datetime.timedelta(days=1)
+
+        while source.last_update + batch_time < timezone.now():
+            last_updated = source.last_update
+
+            until_date = (last_updated + batch_time - one_day).date()
 
-        source.last_update = datetime.datetime.now()
-        source.save()
+            to_update = self.fetch_all_records(
+                filters={'from-update-date':last_updated.date().isoformat(),
+                         'until-update-date':until_date.isoformat()},
+                    cursor=starting_cursor)
+            for record in to_update:
+                try:
+                    bare_paper = self.save_doi_metadata(record)
+                    p = Paper.from_bare(bare_paper)
+                    p.update_index()
+                except ValueError as e:
+                    print((record.get('DOI') or 'unknown DOI') + ': '+unicode(e))
```

Comment:
Should probably be `str` rather than `unicode`, no?

Thinking about a possible migration to Python 3, `unicode` will not be valid. `str` will be the same as `unicode` in Python 3. In Python 2, not sure we need a `unicode` here, but if this is the case, this can be handled with Python-future, see http://python-future.org/compatible_idioms.html#unicode.

Reply:
I didn't know about `from builtins import str`, it sounds great (I thought `six` was needed for this sort of thing).

Reply:
You need the `future` lib to be able to use `builtins` (`pip install future`). Alternative solution is to manage it by hand, such as https://git.weboob.org/weboob/weboob/blob/master/weboob/tools/compat.py#L33-36.

Reply:
I'll take note of this for projects that need to be compatible with both Python 2 and 3. I think I will leave this as it is in our case, as our migration to Python 3 will not be retro-compatible with Python 2 (because we want to migrate to Django 2). So this use of `unicode` should be translated by `2to3` just like the hundreds of others :)